src/export-plan/pdf-generator.ts

```diff
@@ -0,0 +1,361 @@
+            doc.setDrawColor(borderGray);
+            doc.setLineWidth(0.5);
+            // draw middle vertical borders for last row
+            if (data.column.index > 0)
+              doc.line(data.cell.x, data.cell.y + data.row.height, data.cell.x, data.cell.y);
+            // draw top horizontal border for last row
+            doc.line(data.cell.x + data.column.width, data.cell.y, data.cell.x, data.cell.y);
+          }
+
+          tableHeight = data.cell.y + data.row.height - startct + rowHeight;
+
+          if (
+            data.column.index === 2 &&
+            data.section === 'body' &&
+            data.row.index >= 0 &&
+            data.row.index < body.length
+          ) {
+            let yPos = data.cell.y + 3;
+            bubbles[data.row.index].forEach((bubble, index) => {
+              const xPos =
+                data.cell.x + doc.getTextWidth(body[data.row.index][2].split('\n')[index]) + 8;
+              renderBubbles(doc, xPos, yPos, bubble.requirementGroup, bubble.color);
+              yPos += rowFontSize + 1.5;
+            });
+          }
+        },
+      });
+    }
+
+    doc.setDrawColor(borderGray);
+    doc.setLineWidth(0.5);
+    doc.roundedRect(tableX, startct - rowHeight, tableWidth, tableHeight, 4, 4);
+    startct += tableHeight + tableGap;
+  }
+
+  const footerY = doc.internal.pageSize.height - 25;
+  const footerX =
+    (doc.internal.pageSize.width - doc.getTextWidth('downloaded from courseplan.io')) / 2;
+  doc.text('downloaded from ', footerX, footerY, { align: 'left' });
+  doc.textWithLink('courseplan.io', footerX + doc.getTextWidth('downloaded from '), footerY, {
+    url: 'https://courseplan.io',
+  });
+  doc.setDrawColor('#000000');
+  doc.line(
+    footerX + doc.getTextWidth('downloaded from '),
+    footerY + 3,
+    footerX + doc.getTextWidth('downloaded from courseplan.io'),
+    footerY + 3
+  );
+  const pdfName = `${store.state.userName.firstName}'s CoursePlan`;
+  doc.save(pdfName);
+};
+
+// represents a coloured bubble that shows the group a requirement falls in
+type bubble = {
+  requirementGroup: string;
+  color: string;
+};
+type semesterRows = {
+  // the body of the semester table
+  body: string[][];
+  // the list of bubbles for each course in the semester
+  bubbles: bubble[][];
+};
+const getCourseRows = (sem: FirestoreSemester): semesterRows => {
+  const rows: [string[], bubble[]][] = sem.courses
+    .filter((course): course is FirestoreSemesterCourse => !isPlaceholderCourse(course))
+    .map(course => {
+      const [reqs, bubbles] = getFulfilledReqs(course);
+      return [
+        [`${course.code}: ${course.name}`, course.credits.toString(), reqs.join('\n')],
+        bubbles,
+      ];
+    });
+  return { body: rows.map(row => row[0]), bubbles: rows.map(row => row[1]) };
+};
+
+const trimEmptySems = (sems: readonly FirestoreSemester[]): readonly FirestoreSemester[] => {
+  if (sems.length === 0) return [];
```

Comment:
Is this check needed?

Reply:
If length was 0, it'd return sems.slice(0, 0), and I wasn't sure what the behaviour of that was on an empty list, since technically the index 0 doesn't exist, but also 0 >= 0, so maybe it would never even check that. let me know if you'd like me to remove it.

Reply:
You can always test it in the console! `[].slice(0, n)` seems to return `[]` for any `n`, so it's safe to remove.

Reply:
Yep--`a.slice(i, j)` is the sub array from i (inclusive) to j (exclusive).